activesupport/test/log_broadcast_and_tagging_test.rb

```diff
@@ -0,0 +1,137 @@
+    assert_equal("[BMX] Hello\n", @sink2.string)
+  end
+
+  test "tag logs for the whole broadcast without a block" do
+    @broadcast.tagged("BMX").info("Hello")
+
+    assert_equal("[BMX] Hello\n", @sink1.string)
+    assert_equal("[BMX] Hello\n", @sink2.string)
+
+    @sink1.reopen
+    @sink2.reopen
+    @broadcast.info("Hello")
+
+    assert_equal("Hello\n", @sink1.string)
+    assert_equal("Hello\n", @sink2.string)
+  end
+
+  test "tag logs only for one sink" do
+    @logger1.extend(ActiveSupport::TaggedLogging)
+    @logger1.push_tags("BMX")
+
+    @broadcast.info { "Hello" }
+
+    assert_equal("[BMX] Hello\n", @sink1.string)
+    assert_equal("Hello\n", @sink2.string)
+  end
+
+  test "tag logs for multiple sinks" do
+    @logger1.extend(ActiveSupport::TaggedLogging)
+    @logger1.push_tags("BMX")
+
+    @logger2.extend(ActiveSupport::TaggedLogging)
+    @logger2.push_tags("APP")
+
+    @broadcast.info { "Hello" }
+
+    assert_equal("[BMX] Hello\n", @sink1.string)
+    assert_equal("[APP] Hello\n", @sink2.string)
+  end
+
+  test "tag logs for the whole broadcast and extra tags are added to one sink (block version)" do
+    @logger1.extend(ActiveSupport::TaggedLogging)
+    @logger1.push_tags("APP1")
+
+    @logger2.extend(ActiveSupport::TaggedLogging)
+    @logger2.push_tags("APP2")
+
+    @broadcast.tagged("BMX") { @broadcast.info("Hello") }
+
+    assert_equal("[BMX] [APP1] Hello\n", @sink1.string)
+    assert_equal("[BMX] [APP2] Hello\n", @sink2.string)
+  end
+
+  test "tag logs for the whole broadcast and extra tags are added to one sink (non-block version)" do
+    @logger1.extend(ActiveSupport::TaggedLogging)
+    @logger1.push_tags("APP1")
+
+    @logger2.extend(ActiveSupport::TaggedLogging)
+    @logger2.push_tags("APP2")
+
+    @broadcast.tagged("BMX").info("Hello")
+
+    assert_equal("[BMX] [APP1] Hello\n", @sink1.string)
+    assert_equal("[BMX] [APP2] Hello\n", @sink2.string)
+
+    @sink1.reopen
+    @sink2.reopen
+    @broadcast.info("Hello")
+
+    assert_equal("[APP1] Hello\n", @sink1.string)
+    assert_equal("[APP2] Hello\n", @sink2.string)
+  end
+
+  test "can broadcast to another broadcast logger with tagging functionalities" do
+    @sink3 = StringIO.new
+    @sink4 = StringIO.new
+    @logger3 = Logger.new(@sink3, formatter: ActiveSupport::Logger::SimpleFormatter.new)
+    @logger4 = Logger.new(@sink4, formatter: ActiveSupport::Logger::SimpleFormatter.new)
+    @broadcast2 = ActiveSupport::BroadcastLogger.new
+
+    @broadcast2.broadcast_to(@logger3, @logger4)
+    @broadcast.broadcast_to(@broadcast2)
+
+    @broadcast2.push_tags("BROADCAST2")
+
+    @broadcast.tagged("BMX") { @broadcast.info("Hello") }
+
+    assert_equal("[BMX] Hello\n", @sink1.string)
+    assert_equal("[BMX] Hello\n", @sink2.string)
+    assert_equal("[BMX] [BROADCAST2] Hello\n", @sink3.string)
+    assert_equal("[BMX] [BROADCAST2] Hello\n", @sink4.string)
+  end
+
+  test "#silence works while broadcasting to tagged loggers" do
```

Comment:
@davidstosik I saw your comment on https://github.com/rails/rails/issues/49745#issuecomment-1777196805 , but I wasn't able to reproduce the bug.

I wrote this test which is passing, let me know if this is similar to what you had in mind.

Reply:
Yes, that's what I had in mind.
I think I was experiencing issues with `#silence` because I was experimenting with my own implementation of `BroadcastLogger` that did not include `LoggerSilence`.

I wonder, if I have two loggers in my broadcast with different severity levels, will `silence` have an unexpected impact of those loggers' respective level? (Eg. _equalize_ them.)

Reply:
Update: I confirmed that it doesn't look like there is a problem:

```rb
puts "levels= " + logger.broadcasts.map(&:level).join(",")

logger.silence do
  logger.warn("warning")
  logger.info("info")
  puts "in logger.silence"
end

puts "levels= " + logger.broadcasts.map(&:level).join(",")

# Output:
#   levels= 0,2
#   in logger.silence
#   levels= 0,2
# (no other unexpected log)
```

Reply:
@Edouard-chin I'm having more doubts about `#silence` (I guess unrelated to this PR, but as we started the conversation here)



```rb
require "active_support/isolated_execution_state"
require "active_support/logger_silence"
require "active_support/broadcast_logger"

logger = ActiveSupport::BroadcastLogger.new(Logger.new(STDOUT), Logger.new(STDOUT))
logger.silence { logger.warn("Error") };
# W, [2023-10-27T11:14:11.425871 #59954]  WARN -- : Error
# W, [2023-10-27T11:14:11.425960 #59954]  WARN -- : Error
```

Should silencing a `BroadcastLogger` silence _all_ loggers it broadcasts to, or should it silence only the loggers that support silencing?

Reply:
That's a good question. I'd have expect it to silence all logs, no matter if the loggers supports silencing. Worth noting that the behaviour now is the same as before:

```ruby
# Rails 7.0
Rails.logger.extend(ActiveSupport::Logger.broadcast(Logger.new(STDOUT))

Rails.logger.silence { Rails.logger.info("hello") }
# I, [2023-10-27T14:00:18.659473 #86197]  INFO -- : hello
```

Maybe we can change this behaviour in another PR and silence the whole broadcast. What do you think?

Reply:
I'm definitely not asking it to be done in this PR! 👍🏻 
When it comes to PRs, the smaller the better. 👌🏻

Reply:
Ah yeah no worries! I was just wondering if silencing the whole broadcast is the behaviour you would expect and whether we should change it.

Reply:
Yes, that would be my own expectation too. `BroadcastLogger` is trying to be "transparent" to the user, so doing `Rails.logger.silence` should, without surprise, stop `Rails.logger.log` from outputting anything on all channels.

Reply:
I'll mark this as resolved to tidy up this PR, we can address this in a follow up as discussed. Let me know if you have other questions or concerns though :)